New module: Add amazon cloudfront distribution module. (cloud/amazon/cloudfront_distribution)

wilvk opened the pull request:
#### SUMMARY

This module allows the following cloudfront actions:

- create distribution
- update distribution
- delete distribution
- create streaming distribution
- update streaming distribution
- delete streaming distribution
- tag a resource
- untag a resource
- generate a pre-signed cloudfront url from a PEM private key

I saw this as a gap item as the `cloudfront_facts.py` module already exists but not `cloudfront.py`.

Each action has a set of valid parameters that an be used in conjunction. These are illustrated in the examples section.

This PR also renames the class for `cloudfront_facts.py` as `CloudfrontFactsManager` as `CloudfrontManager` is more appropriate for this module.

#### ISSUE TYPE

New Module Pull Request

#### COMPONENT NAME

./lib/ansible/modules/cloud/amazon/cloudfront.py

#### ANSIBLE VERSION

```
ansible 2.3.0 (devel 652b4d220a) last updated 2017/04/30 23:12:22 (GMT +1100)
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.10 (default, Feb  6 2017, 23:53:20) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```

#### ADDITIONAL INFORMATION

This PR is migrated from PR #24151 as the branch and PR comment history was irreparable and has thus been closed.

comments from @willthames:

`create_streaming_distribution` and `delete_streaming_distribution` and the other similarly named parameters are of concern - we should be stating the desired state of `streaming_distribution` (present or absent).

This feels like it should probably be multiple modules.

The use of `tag_resource` and `untag_resource` is problematic - just provide a `tags:` parameter and update tags accordingly

comments from @wilvk:

Thanks for the feedback @willthames ! I'll get onto those changes:

consolidate tag and untag into just `tags:`
use `status` parameter instead of `create_distribution` `delete_distribution` `update_distribution` `duplicate_distribution` `validate_distribution` etc.
could possibly be multiple modules but not sure how it should be split - distribution and streaming_distribution have dependencies/common methods. Am open to suggestions, however..

wilvk committed:
added cloudfont.py, modified cloudfront_facts.py class name and fixed a minor bug

wilvk committed:
first pass fixing documentation compliance

wilvk committed:
pep8 styled code

wilvk committed:
fixes for pep8

wilvk committed:
fixes for shippable

wilvk committed:
fixes for shippable

wilvk committed:
fixes for shippable

wilvk committed:
fixes for shippable
removed tag and untag resource and replaced with just tags:
pep8 fixes
added purge_tags option

wimnat commented:
I don't see how this module can be idempotent as there's no state parameter. It's more procedural which is not really the ansible way. 

You should have a purge_tags option if you're supporting tags to decide what to do with existing tags that aren't defined in the playbook.

wilvk committed:
changed boolean parameters for actions into resource: and state: tags for idempotence

Commit:
fix for doc

Commit:
fixes for create from tags implementation

wilvk committed:
documentation, fixed default for enabled, renamed a variable

wilvk committed:
shipit fixes

wilvk commented:
@willthames @wimnat ready_for_review !

Changes include:
- changed `tag_resource` and `untag_resource` to just `tags` and a `purge_tags` parameters.
- changed all create update delete duplicate verify action parameters to just 2 parameters:
    - resource
    - state
where resource is one of `distribution`, `streaming_distribution`, `invalidation` and `origin_access_identity`
and state is once of `present`, `absent`, `updated`, `duplicated` and `verified`
- added more thorough documentation.
- various minor fixes

shawnsi commented:
As is the `updated` state is not very reliable due to inconsistencies in required parameters between the boto3 `create_distribution` and `update_distribution` calls.  I've documented the problem at https://github.com/boto/boto3/issues/1083 upstream.

When trying to use `updated` to manage a distribution without specifying all parameters (even non-required ones) errors like this bubble up:

```
TASK [create cloudfront distribution] *******************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error w
as: IllegalUpdate: An error occurred (IllegalUpdate) when calling the UpdateDistribution oper
ation: OriginReadTimeout is required for updates.
```

If this is in fact an upstream AWS requirement in the `UpdateDistribution` API this module will need to enforce adding those parameters.  That would either require:

1) Extending the default value management in this module to non-required parameters.
2) Merging the current configuration with the module parameters.  This would be ideal but is not trivial to implement.

@wilvk any thoughts about how to proceed here?

shawnsi commented:
@wilvk I also have a small fix for missing `quantity` fields at https://github.com/meredith-digops/ansible/commit/7898b4724a89e15706072183bad47d88b101556b.  I can't seem to open a PR against your fork.  Can you take a look and lift that patch if it makes sense to you?

wilvk committed:
fix for missing quantity and fix for validate_attribute_with_allowed_values for lists, added to examples

wilvk committed:
removed invalid method combo from examples

wilvk committed:
removed unused variable

wilvk committed:
pep8 fixes

wilvk committed:
pep8 fixes

shawnsi commented on lib/ansible/modules/cloud/amazon/cloudfront.py:
```diff
@@ -0,0 +1,1765 @@
+        self.__default_cache_behavior_smooth_streaming = False
+        self.__default_cache_behavior_forwarded_values_forward_cookies = 'none'
+        self.__default_cache_behavior_forwarded_values_query_string = True
+        self.__default_trusted_signers_enabled = True
```
Shouldn't this default to `False` like in line 896?  This is causing distributions to come up with trusted signers enabled but no trusted signers configured by default.

wilvk committed:
pep8 fixes

wilvk committed:
default trusted signers to false

wilvk commented:
Thanks for the feedback @shawnsi ! I've implemented your two suggestions:
- add quantity to `allowed_methods` and `cached_methods` - done slightly different as there is already a method to do the quantity for an aws list
- set `default_trusted_signers` to false - for consistency and logic

WRT OriginReadTimeOut - I didn't notice that attribute as an issue specifically, but I did notice the inconsistencies you mention with update vs create and have catered for them where I encountered them. I think there was only one instance where I had to make concessions to 'back-fill' with defaults for this circumstance.

shawnsi commented:
@wilvk do you have examples of accommodating the additional required values for the `update_distribution` calls?

wilvk commented:
@shawnsi unfortunately not - from memory it was a minor fix, like an attribute name had to be present even though there was nothing in it - so I just set to `foo = ''` or `foo = {}`. I vaguely remember there was an SO question highlighting the same issue.

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront.py:
```diff
@@ -0,0 +1,1765 @@
+author: Willem van Ketwich (@wilvk)
+
+options:
+    resource:
```
I think this is out of date, because I see the argspec has a `state` parameter.

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront.py:
```diff
@@ -0,0 +1,1765 @@
+                config['web_a_c_l_id'] = web_acl_id
+            return config
+        except Exception as e:
+            self.module.fail_json(msg="error validating distribution config parameters - " + s
```
I think supporting multiple different resources in one module isn't a sustainable idea. That means we need a test matrix of 4 resource types times all the parameters for all 4. Would it be possible to split these into multiple modules? Looking at the CF API, it looks like we'd want to split:

- "cloudfront_distribution" which would support streaming/non-streaming distributions
- "cloudfront_invalidation" which would be exactly what it sounds like
- "cloudfront_origin_access_identity" which would output the proper S3 canonical user ID. 

Splitting like that would let the cloudfront_distribution module drop at least 8 parameters based on a cursory count.

wilvk replied:
Thanks for the feedback @ryansb ! That seems like a sensible way to split the module. I'll get onto that.

wilvk committed:
split to three separate modules - cloudfront_distribution, cloudfront_invalidation and cloudfront_origin_access_identity

wilvk committed:
shippable fixes

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
doc fix

wilvk committed:
fixed default, naming of method and error message

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,1608 @@
+          updated - updates an existing distribution
+          duplicated - duplicates an existing distribution
+              (can also make updates at the same time)
+          verified - verifies a configuration parameters
```
It looks like "validated" is what's actually used in the module. I'm not entirely sure what this option is here for - wouldn't the normal way to validate a config be to use the facts module + assert?

wilvk replied:
verified is more like a dry-run without actually creating the distribution so the user can test their configuration is correct in an ad-hoc manner before creating it. I thought this would be useful due to the complexity of the distribution object, but yes, you are correct - facts + assert would be the correct way to validate the config after creation.

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,1608 @@
+          present - creates a new distribution
+          absent - deletes an existing distribution
+          updated - updates an existing distribution
+          duplicated - duplicates an existing distribution
```
I'm not a fan of this option. It's more of a command-oriented setup than idempotent. What if I run this twice on a distribution, creating CNAME conflicts with the duplicate made during the first run?

wilvk replied:
No worries - will remove. The idea was if a user has a base distribution with all their finely configured cache behaviours and want to duplicate it for a new website in the creation process, this would automate that. However, if they really want to duplicate a distribution they can just get the config from cloudfront_facts and use that as a basis anyway.

ryansb replied:
Yeah - I get the idea of "take a config and use it" but I think that's a touchy enough project that I'd rather have users transfer stuff either through facts or manually, since I can see a lot of weird edge cases being a problem with a "duplicate" command.

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,1608 @@
+        - The desired state of the distribution or streaming distribution.
+          present - creates a new distribution
+          absent - deletes an existing distribution
+          updated - updates an existing distribution
```
Generally updated isn't a state we use, because that implies that "present" would *not* update the config to converge on what the module specifies.

wilvk replied:
No worries - will amend so that present checks for the presence of the distribution and updates, otherwise creates the distribution.

ryansb commented:
Also, since these modules are separate, can you move them to distinct PRs so we can review/merge them independently?

wilvk committed:
invalidation paraneter validation

wilvk commented:
@ryansb ready_for_review !

I've split the module into the 3 modules as advised:
- cloudfront_distribution
- cloudfront_invalidation
- cloudfront_origin_access_identity

Have also added:
- extra validation for `state` in `cloudfront_distribution` to determine whether performing an action or generating a presigned url
- `cloudfront_invalidation` generates `distribution_id` if `alias` is provided
  - relevant parameter checking for `distribution_id` and `alias`

wilvk committed:
removed invalidation and origin access identity and placed in separate PRs for separate reviews

wilvk committed:
removed duplicate and validate, merged updated with present, refactoring of variables

wilvk committed:
refactoring - pep8 compliant length, validation for methods

wilvk committed:
more refactoring for pep8, smaller functions and minor fixes for null objects

wilvk commented:
@ryansb have made the changes as advised:
- merged `state:updated` into `state:present`
- removed `state:duplicate`
- removed `state:verified`
- split into separate PRs

Also done:
- added function for determining `distribution_id` from `alias` or `aliases` as part of `state:present` merge
- refactoring some of the larger functions into smaller ones
- conform to PEP8 for 90 character lines where possible
- better validation of `allowed_methods` and `cached_methods`
- added `traceback` to try-catch blocks

wilvk committed:
fixed typos, formatting, added mutually exclusive parameters

wilvk committed:
Update cloudfront_distribution.py

wilvk committed:
moved helpers to separate module, formatting

wilvk committed:
fixes for idempotent distribution using caller_reference

wilvk committed:
formatting for pep8 79 characters

wilvk committed:
doc fixes

wilvk committed:
doc mods

wilvk committed:
helpers as modules, doc fixes, doc pricing info, formatting for pep8

wilvk committed:
added object to classes

wilvk committed:
removed unused method

wilvk commented:
@ryansb I've made a few more changes if that's ok, including:
- added method for getting `(streaming)distribution_id` from `caller_reference` as this is also a valid identifier and can determine whether the distribution is created or updated (if included).
- fixed/cleaned up documentation
- added defaults for custom origin `ReadTimeout` and `KeepaliveTimeout`as these values are required for updates but not create calls (as mentioned earlier by @shawnsi).
- more formatting for pep8

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,1936 @@
+      description:
+        - The name of an alias that is used in a distribution. This is used to
+          effectively reference a distribution by its alias as an alias can
+          only be used by one distribution per AWS account. This variable
```
I think aliases are unique across AWS - so there's only one CF distribution allowed to have the fake.site.com alias at any given time. 

> You cannot add an alternate domain name to a CloudFront distribution if the alternate domain name already exists in another CloudFront distribution, even if your AWS account owns the other distribution.

wilvk replied:
This comment is showing as outdated - is this still an issue?

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,1936 @@
+                config = {}
+            if aliases is not None:
+                config['aliases'] = helpers.python_list_to_aws_list(aliases)
+  
```
I like that you've put some reusable functions in module_utils.

I'm still reviewing this, but one thing I've noticed is that this module always shows changed=True and e_tag is always modified.
When I run
```
    - name: try one
      cloudfront_distribution:
        state: present
        enabled: false
        e_tag: "E3OUIFFZ9OWBTR"
```
Running that once does modify the expected distribution. But running a second time, a new distribution is created and e_tag disregarded (since with the first run e_tag is modified regardless of whether or not changes should be made). That seems unexpected.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,1936 @@
+                config = {}
+            if aliases is not None:
+                config['aliases'] = helpers.python_list_to_aws_list(aliases)
+  
```
I like that you've put some reusable functions in module_utils.

I'm still reviewing this, but one thing I've noticed is that this module always shows changed=True and e_tag is always modified.
When I run
```
    - name: try one
      cloudfront_distribution:
        state: present
        enabled: false
        e_tag: "E3OUIFFZ9OWBTR"
```
Running that once does modify the expected distribution. But running a second time, a new distribution is created and e_tag disregarded (since with the first run e_tag is modified regardless of whether or not changes should be made). That seems unexpected.

wilvk replied:
Thanks for the feedback @s-hertel ! Yes, that does seem unexpected. However, I'm unable to run that as a task as either a distribution_id, caller_reference or an alias is usually required for updating a distribution and either default_origin_domain_name or origins[] are required for a new distribution. Are there extra parameters you are running this task with? The playbook I'm using is:

```
- hosts: 127.0.0.1
  connection: local
  tasks:
    - cloudfront_distribution:
        state: present
        enabled: false
        e_tag: "E1I5YC9H172ZQX"
```

(after fixing a minor bug) the output I'm getting is:

` "msg": "both origins[] and default_origin_domain_name have not been specified. please specify at least one."`

wilvk committed:
added default origins array

wilvk committed:
documentation fixes

wilvk committed:
renamed variable for clarity

wilvk committed:
pep8 fixes

ryansb reviewed (commented):
When using this, it seems to always exit with "changed=true" when the distribution is in the InProgress state, even when changing nothing. 

Reproduction: 

```
---
- hosts: localhost
  gather_facts: no
  connection: local
  tasks:
    - cloudfront_distribution:
        state: present
        distribution_id: # filled in on the second run after the distribution exists
        default_origin_domain_name: anywhere.com
        comment: modified by ansible cloudfront.py
```

I don't see where that would cause the changed state, but since it isn't altering any settings that seems like the wrong behavior.

wilvk committed:
fixed distribution_ids bug

wilvk committed:
doc fixes

wilvk commented:
Hi @ryansb I think what is happening when that playbook is run twice is that it is actually creating two cloudfront distributions. As the `default_domain_name_origin` is not a unique identifier, it is not found in the list of distributions and thus creates a new distribution each time. The `default_domain_name_origin` is what the distribution looks at to get the data for the distribution and can be identical over multiple distributions. Each alias in `aliases[]`, however, cannot be identical and must be unique over multiple distributions as it is what is used by the end-user to access the distribution (if a cloudfront.net domain name is not used). 

As the `caller_reference` value is not set, there is nothing to identify the distribution as unique from the playbook. If the `caller_reference` is not set, it is generated internally in the module as a datetime stamp, (for example: "caller_reference": "2017-05-31T22:05:49.067767"). Using the following playbook with a `caller_reference` set still produces `changed=true`.

```
---
- hosts: localhost
  gather_facts: no
  connection: local
  tasks:
    - cloudfront_distribution:
        state: present
        default_origin_domain_name: anywhere.com
        caller_reference: unique distribution
        comment: modified by ansible cloudfront.py
```

Looking into this and running the playbook with `-vvv` I can see that the `e_tag` and the `last_modified_time` fields have actually changed even though nothing in the distribution has changed. This must be expected behaviour by AWS when a request to update a distribution has been made from the API. I could do a comparison of the distribution config before and after the API call and if there are no changes to set `changed=false`, but even this may be incorrect as at a minimum, the last modified time and `e_tag` values have changed. Alternatively, I could compare the current distribution config to the proposed changes and if identical, skip the update and set `changed=false`, but this may be a bit excessive and complicated to implement. Let me know what you think.


Testing this I identified a bug where it was checking aliases as distribution_ids and have fixed accordingly.

ryansb commented:
But how is it not matching on distribution_id then? That's unique, so if I specify the ID the second time (with the ID returned by the first) it should show unchanged, right?

wilvk commented:
Ah, I see what you're saying now, my apologies - you've added the distribution_id into the play the second time and it is still showing as changed. Yes, that is matching the distribution and attempting to update it, the same as if the caller_reference were used (in my example above) and causing the same result - the e_tag and last modified time are still updated..

wilvk committed:
shippable

wilvk commented:
@ryansb can you confirm that it should still be `changed=no` with a changed `e_tag` and `last_modified_date` and I will start working on a fix for this.

wilvk committed:
fix for changed=no when there are no differences between proposed and existing distribution config, added to doc for caller reference

wilvk commented:
Hi @ryansb I've implemented the `changed=no` check for testing whether an update is required. It compares the existing to the proposed configuration and does an update if they are different. 
Also done:
- fixed a bug with streaming distributions
- updated caller_reference documentation
- refactored code for updating tags into separate service_mgr method

wilvk committed:
minor doc fixes

wilvk committed:
mutually exclusive parameters, refactoring, bug fix

wilvk committed:
fixed doc for mutually exclusive params

wilvk commented:
@ryansb I finally figured out why it was always showing as `changed=yes` for the scenario you presented - it is because using `default_origin_domain_name` was originally intended for a quick 'bootstrapping' of a distribution without having to fill in all the details and this consequentially creates a new origin with defaults each time when this parameter is included. 

On a second and subsequent runs, it was filling in the same details for `origins[]`, only with a new `target_origin_id` that defaults to a timestamp. As this timestamp is always different between runs, it was appearing as a new cloudfront distribution each time.

I have made the parameters [ [`default_origin_domain_name`] and [`distribution_id`, `alias`, `caller_reference`] ] mutually exclusive as well as the respective parameters for streaming distributions to prevent the user from performing this unintended behaviour.

ryansb commented:
But how is it not matching on distribution_id then? That's unique, so if I specify the ID the second time (with the ID returned by the first) it should show unchanged, right?

wilvk commented:
@ryansb Yes, it should return unchanged on the second run, and it does. Unless I'm missing something, this should be fixed now.

Below shows the exact steps I am using to test this scenario. Please let me know if this is not correct. Also, I am sure you are, but please make sure you're using the latest code.

example below:

**first run:**

command: `ansible-playbook test_dis1.yml -vvvv`

playbook `test_dis1.yml`:

```
---
- hosts: localhost
  gather_facts: no
  connection: local
  tasks:
    - cloudfront_distribution:
        state: present
#        distribution_id: # filled in on the second run after the distribution exists
        default_origin_domain_name: anywhere.com
        comment: modified by ansible cloudfront.py

```
truncated result:

> ...
>  "location": "https://cloudfront.amazonaws.com/2017-03-25/distribution/E27XX2AGOSGPMR"
> }
> META: ran handlers
> META: ran handlers
> 
> PLAY RECAP *************************************************************************************************************
> localhost                  : ok=1    changed=1    unreachable=0    failed=0

**second run:**

command: `ansible-playbook test_dis2.yml -vvvv`

playbook `test_dis2.yml`:

```
---
- hosts: localhost
  gather_facts: no
  connection: local
  tasks:
    - cloudfront_distribution:
        state: present
        distribution_id: E27XX2AGOSGPMR
  #        default_origin_domain_name: anywhere.com
        comment: modified by ansible cloudfront.py
```

truncated result:

> ...
>            "validate_certs": true,
>             "viewer_certificate": null,
>             "web_acl_id": null
>         }
>     }
> }
> META: ran handlers
> META: ran handlers
> 
> PLAY RECAP *************************************************************************************************************
> localhost                  : ok=1    changed=0    unreachable=0    failed=0

wilvk committed:
fix for multiple runs with caller_reference - not does not show changed on second and subsequent runs - set all defaults to None when updating as defaults only valid for bootstrapping, removed superfluous code, refactored bool vars.

wilvk committed:
add missed default to exclude from update

wilvk commented:
@ryansb I've made a change to allow the `default_domain_name_origin` parameter to be used in conjunction with the `caller_reference` parameter so that the following play will run. 

```
---
- hosts: localhost
  gather_facts: no
  connection: local
  tasks:
    - cloudfront_distribution:
        state: present
        default_origin_domain_name: anywhere.com
        caller_reference: unique distribution
        comment: modified by ansible cloudfront.py
```

I've also set all `default_` parameters not to be used on second and subsequent runs to ensure the play is idempotent when using these parameters.

@ryansb, @s-hertel - please let me know if there is any other feedback for this module that needs actioning.

Thanks,

wimnat commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2039 @@
+            distribution.
+          absent - deletes an existing distribution.
+      choices: ['present', 'absent']
+      required: false
```
State should be required

wilvk replied:
Thanks for the feedback - there is extra functionality in this module for generating presigned urls that prevents the `state `parameter from being required - ie. using `generate_presigned_url` (see example on line 577).

wimnat commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2039 @@
+          if specified.
+      required: false
+      default: 'no'
+      choices: ['yes', 'no']
```
Most other modules default to yes so I would do the same here

wilvk replied:
I appreciate the feedback, but I'm not sure that's an entirely safe default to have - any modifications to a distribution are going to purge all tags regardless of whether tags have been specified or not. This would happen even for a small task to modify an existing distribution, where it would unwittingly remove all existing tags. 

We could rename it to `keep_tags` and have a default of yes if that would satisfy the default=yes requirement...

wimnat commented:
Usually if no tags are specified, nothing happens. An empty list is
required to purge all tags.

The reason I believe that it's usually default yes is because ansible,
first and foremost, is meant to be enforcing desired state.

It's not imperative,I just thought I'd point out how other modules are
doing it for consistencies sake.

On 19 Jun. 2017 8:39 pm, "Willem van Ketwich" <notifications@github.com>
wrote:

*@wilvk* commented on this pull request.
------------------------------

In lib/ansible/modules/cloud/amazon/cloudfront_distribution.py
<https://github.com/ansible/ansible/pull/24292#discussion_r122673694>:

> +        - Used for distributions and streaming distributions.
+          Should be input as a I(list[]) of I(Key) I(Value) pairs.
+      required: false
+
+    purge_tags:
+      description:
+        - Specifies whether existing tags will be removed before adding
+          new tags.
+          When C(purge_tags=yes), existing tags are removed and C(tags) are
+          added, if specified. If no tags are specified, it removes all
+          existing tags for the distribution.
+          When C(purge_tags=no), existing tags are kept and C(tags) are added,
+          if specified.
+      required: false
+      default: 'no'
+      choices: ['yes', 'no']

I appreciate the feedback, but I'm not sure that's an entirely safe default
to have - any modifications to a distribution are going to purge all tags
regardless of whether tags have been specified or not. This would happen
even for a small task to modify an existing distribution, where it would
unwittingly remove all existing tags.

We could rename it to keep_tags and have a default of yes if that would
satisfy the default=yes requirement...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ansible/ansible/pull/24292#discussion_r122673694>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AHFORuhZH78arWE2MPc2VdULmu0v6H1fks5sFk_1gaJpZM4NQmQ1>
.

wilvk commented:
@ryansb @s-hertel @wimnat just thinking, would a wait option be useful for this module to poll until cloudfront changes are completed? This would save the user having to check the CF status before doing an action based on it's state - eg. doing Selenium tests after changes have propagated.

wilvk committed:
added wait_for_processing parameter, added validation on delete to make repeat delete tasks idempotent, refactored list of elif statements

wilvk commented:
I've added the following changes:
- `wait_for_processing` parameter (defaults to true) to poll the distribution until the status becomes 'Deployed'
- fixed idempotence on deleting a distribution - previously it would error if the distribution wasn't present on second and subsequent runs.
- refactoring of unnecessary `elif` statements.

wilvk committed:
pep8 fix

wilvk commented:
ready_for_review

willthames commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+    return pascalize(snake_dict)
+
+
+def pascal_dict_to_snake_dict(pascal_dict, split_caps=False):
```
camel_dict_to_snake_dict already exists in module_utils/ec2.py

wilvk replied:
see my comments above. Additionally, `split_caps` is used for processing intermediate variable names where there is not a direct conversion possible.

e.g. `LambdaFunctionARN` -> `lambda_function_a_r_n` would otherwise be `lambda_function_arn` which does not convert back to `LambdaFunctionARN` but to`LambdaFunctionArn` which is incorrect for boto3. I hope I've explained this correctly.

willthames commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+    return dictionary
+
+
+def snake_dict_to_pascal_dict(snake_dict):
```
snake_dict_to_camel_dict already exists in module_utils/ec2.py

wilvk replied:
`snake_dict_to_camel_dict` is different from `snake_dict_to_pascal_dict` in that pascal has the first letter capitalised, whereas camel has first letter lower-cased. This is an important distinction as this function is used for formatting parameters passed to boto3 that are case-sensitive. If it were just the first-level items of the dict, I could use the `snake_dict_to_camel_dict` function with `.capitalize()` but there are objects that have nested objects that have nested objects... and so on.

willthames commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+      required: false
+      default: aws defaults this to 'http2'
+
+    is_ipv6_enabled:
```
Should just be `ipv6_enabled`

wilvk replied:
replaced all instances with ipv6_enabled

willthames commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+          Valid for both distributions and streaming distributions.
+      choices: ['yes', 'no']
+      default: 'no'
+    wait_until_processed:
```
Use `wait`, it's standard for similar modules. Requires a configurable `wait_timeout`.

I'm not sure whether it's worse to have a default `wait` with a timeout in excess of 20 minutes so that it typically succeeds, or just no wait so that people at least knowingly set a long wait time. I'd think default to waiting off, but default to 30 minutes wait_timeout if people do set `wait`

wilvk replied:
Done. Using seconds for `wait_timeout` so set to 1800s (30 mins). Also added to docs.

willthames commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+                msg="error validating tagging parameters - " + str(e) +
+                "\n" + traceback.format_exc())
+
+ 
```
This module is not idempotent if you can't run it twice with the same arguments. 

If you have to know the distribution ID to update it, but not create it, then this is not going to work as it should. Really it should be searching for a distribution that matches the unique distribution properties (presumably the origins of the distribution)

wilvk replied:
`distribution_id` is not necessary. can use `alias` or `caller_reference` parameters for idempotence.

willthames commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+                msg="error validating tagging parameters - " + str(e) +
+                "\n" + traceback.format_exc())
+
+ 
```
These shouldn't be mutually exclusive - if they're passed in with state=present then a presigned url should be returned.

wilvk replied:
not sure what parameters you are referring to? 
I'm guessing `['state', 'generate_presigned_url'] `so have removed.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+    distribution_id:
+      description:
+        - The id of the cloudfront distribution.
+          This parameter can be exchanged with C(alias) or C(caller_reference)
```
I think I() is for option and C() is for file. C() is used throughout this documentation.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+            distribution.
+          absent - deletes an existing distribution.
+      choices: ['present', 'absent']
+      required: false
```
Not necessary to document required: false for any option as it is the default.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+    generate_presigned_url:
+      description:
+        - Generates a presigned url for a distribution from a private
+          I(.pem) file key.
```
Maybe C(.pem)? I'm not sure you need to specify that .pem is a file type since it's just the extension. @gundalow ?

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+          This parameter can be used instead of C(distribution_id) or
+          C(streaming_distribution_id) to reference an existing distribution.
+          If not specified, this defaults to a datetime stamp of the format
+            'YYYY-MM-DDTHH:MM:SS.ffffff'
```
extra spaces

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+
+    pem_key_path:
+      description:
+        - The path on the host to the I(.pem) private key file.
```
Same as comment on I(.pem) above.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+
+    presigned_url_expire_date:
+      description:
+        - The expiry date of the presigned url. Date format is I(YYYY-MM-DD)
```
YYYY-MM-DD isn't a module option, so I think you can remove the I and parentheses

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+    tags:
+      description:
+        - Used for distributions and streaming distributions.
+          Should be input as a I(list[]) of I(Key) I(Value) pairs.
```
I think you can just say '...input as a list of key value pairs' since these are not module options.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+      description:
+        - A config element that specifies the path to request when the user
+          requests the origin.
+          eg. if specified as 'index.html', this maps to
```
e.g.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+        - A config element that specifies the path to request when the user
+          requests the origin.
+          eg. if specified as 'index.html', this maps to
+          www.example.com/index.html when www.example.com is called by the user.
```
'U(www.example.com/index.html) when U(www.example.com)...'

s-hertel commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+                config_node_items = config_node.items()
+            else:
+                config_node_items = []
+            config[node_name] = dict(config_node_items + validated_node.items())
```
I don't think this will work with python3

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+                msg="error validating tagging parameters - " + str(e) +
+                "\n" + traceback.format_exc())
+
+ 
```
There should be a check before this that HAS_BOTO3 is true.

s-hertel replied:
There should be a check before this that HAS_BOTO3 is true.

s-hertel replied:
If the tags option is a dict you can use python_dict_to_boto3_tag_list instead of implementing the functionality here.

s-hertel replied:
Changed = True unless trying to delete something that doesn't exist isn't idempotent. Do you think it's possible to make this idempotent?

s-hertel replied:
There should be a check before this that HAS_BOTO3 is true.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict
+from ansible.module_utils.basic import AnsibleModule
+from ansible.modules.cloud.amazon.cloudfront_facts import (
+    CloudFrontFactsServiceManager)
```
While this works, I think we try to avoid importing things from other modules in general. Maybe it doesn't pertain to _facts modules. (?)

s-hertel commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
```
Since module_utils/aws/ has been merged this could be moved to module_utils/aws/cloudfront.py now

s-hertel commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+        if isinstance(validated_node, dict):
+            config_node = config.get(node_name)
+            if config_node is not None:
+                config_node_items = config_node.items()
```
Since in the else this is defaulting to an empty list I think this line should be `config_node_items = list(config_node.items())` to be python3 compatible

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+        try:
+            changed = False
+            if purge_tags:
+                self.remove_all_tags_from_resource(arn)
```
purge_tags doesn't need to remove all tags - you can use compare_aws_tags() (in module_utils/ec2) and then call untag_resource() with the list of tag_keys that are returned.

s-hertel replied:
purge_tags doesn't need to remove all tags - you can use compare_aws_tags() (in module_utils/ec2) and then call untag_resource() with the list of tag_keys that are returned.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+                msg="error validating and updating tags - " + str(e) + "\n" +
+                traceback.format_exc())
+
+    def paginated_response(self, func, result_key=''):
```
Is this doing something differently than https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py#L320?

s-hertel replied:
Is this doing something differently than https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py#L320 besides getting rid of the ResponseMetadata?

s-hertel commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
```
Since module_utils/aws/ has been merged this could be moved to module_utils/aws/cloudfront.py now

wilvk replied:
will do this as a final thing as am afraid I'll break something in my branch doing a pull from devel (being somewhat of a a git n00b and all)

s-hertel commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+        if isinstance(validated_node, dict):
+            config_node = config.get(node_name)
+            if config_node is not None:
+                config_node_items = config_node.items()
```
Since in the else this is defaulting to an empty list I think this line should be config_node_items = list(config_node.items()) to be python3 compatible and always get a list.

wilvk replied:
done

s-hertel commented on lib/ansible/module_utils/cloudfront.py:
```diff
@@ -0,0 +1,122 @@
+                config_node_items = config_node.items()
+            else:
+                config_node_items = []
+            config[node_name] = dict(config_node_items + validated_node.items())
```
config_node_items may be an iterator or empty list right now in python3. I don't think merging a list and iterator will work. Might need list(validated_node.items()).

wilvk replied:
done

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+            distribution.
+          absent - deletes an existing distribution.
+      choices: ['present', 'absent']
+      required: false
```
Not necessary to document required: false for any option as it is the default. Can update all of the following documentation to reflect this.

wilvk replied:
done

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+    distribution_id:
+      description:
+        - The id of the cloudfront distribution.
+          This parameter can be exchanged with C(alias) or C(caller_reference)
```
I think I() is for option and C() is for file. C() is used throughout this documentation.

wilvk replied:
done

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+    generate_presigned_url:
+      description:
+        - Generates a presigned url for a distribution from a private
+          I(.pem) file key.
```
Maybe C(.pem)? I'm not sure you need to specify that .pem is a file type since it's just the extension. @gundalow might know.

wilvk replied:
done

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+          This parameter can be used instead of C(distribution_id) or
+          C(streaming_distribution_id) to reference an existing distribution.
+          If not specified, this defaults to a datetime stamp of the format
+            'YYYY-MM-DDTHH:MM:SS.ffffff'
```
extra spaces

wilvk replied:
fixed

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+
+    pem_key_path:
+      description:
+        - The path on the host to the I(.pem) private key file.
```
Same as comment on I(.pem) above.

wilvk replied:
done

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+
+    presigned_url_expire_date:
+      description:
+        - The expiry date of the presigned url. Date format is I(YYYY-MM-DD)
```
YYYY-MM-DD isn't a module option, so I think you can remove the I and parentheses

wilvk replied:
done

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+    tags:
+      description:
+        - Used for distributions and streaming distributions.
+          Should be input as a I(list[]) of I(Key) I(Value) pairs.
```
I think you can just say '...input as a list of key value pairs' since these are not module options.
But can tags just be a dict? You could then use ansible_dict_to_boto3_tag_list(). It feels like some functionality is being duplicated.

wilvk replied:
done, however `ansible_dict_to_boto3_tag_list()` doesn't handle int values correctly for a boto3 list - they should be converted to strings but are not, causing boto3 to throw an error. e.g.

```
- hosts: 127.0.0.1
  connection: local
  tasks:
  - cloudfront_distribution:
      state: present
      default_origin_domain_name: www.my-cloudfront-origin.com
      caller_reference: test cfnt distribution
      comment: test comment
      tags:
        Name: example distribution
        Project: example project
        Priority: 111
```

causes the error:

```
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter DistributionConfigWithTags.Tags.Items[1].Value, value: 111, type: <type 'int'>, valid types: <type 'basestring'>
```

Have added to the tags doc for this.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+      description:
+        - A config element that specifies the path to request when the user
+          requests the origin.
+          eg. if specified as 'index.html', this maps to
```
e.g.

wilvk replied:
done for 2 instances

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict
+from ansible.module_utils.basic import AnsibleModule
+from ansible.modules.cloud.amazon.cloudfront_facts import (
+    CloudFrontFactsServiceManager)
```
While this works, I think we try to avoid importing things from other modules in general. Maybe it doesn't pertain to _facts modules. (?)

wilvk replied:
This isn't trivial to remove as there are 18 instances of `cloudfront_facts` methods being called and would result in duplicate code between the two modules. It looks like shippable is failing on importing this module too now. If it's a show-stopper, I can implement the required functions but would prefer not to if I can avoid it.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+                    custom_origin_config['origin_keepalive_timeout'] = (
+                        self.__default_custom_origin_keepalive_timeout)
+                if 'http_port' not in custom_origin_config:
+                    custom_origin_config['h_t_t_p_port'] = (
```
You don't need line breaks for lines less than 160 characters. Noticing a lot of these.

wilvk replied:
fixed for all occurrences, thought that the pep8 limit was 79 chars for some reason

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+            cache_behavior['trusted_signers'] = self.validate_trusted_signers(
+                cache_behavior.get('trusted_signers'))
+            return cache_behavior
+        except Exception as e:
```
It's not ideal to have such long try/except blocks (or to catch Exception). I'm not sure if there's a more manageable way to do it?

wilvk replied:
refactored this function into another function and removed the try/except block so all errors are caught further down now. let me know if this is not what you meant.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+                msg="error validating tagging parameters - " + str(e) +
+                "\n" + traceback.format_exc())
+
+ 
```
You don't need to specify required=False or default=None or type='str' for any of these as it is the default.

wilvk replied:
fixed

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -0,0 +1,2106 @@
+                msg="error validating tagging parameters - " + str(e) +
+                "\n" + traceback.format_exc())
+
+ 
```
There should be a check before https://github.com/ansible/ansible/pull/24292/files#diff-aaef1b3abde00153e8695bfdf79c5e55R1882 that HAS_BOTO3 is True.

wilvk replied:
added check

wilvk committed:
renamed all instances of is_ipv6_enabled to ipv6_enabled

wilvk committed:
changed param to wait and added wait_timeout. defaults to 1800s (30 min). added to docs

wilvk committed:
removed mutually exclusive

wilvk committed:
added list for python3

wilvk committed:
added list() for python 3

wilvk committed:
removed all required: false, added default for wait_timeout to docs

wilvk committed:
changed to C(.pem) as advised

wilvk committed:
removed extra spaces and wrapping

wilvk committed:
removed I()

wilvk committed:
eg. to e.g.

wilvk committed:
changed all C() to I() for parameters

wilvk committed:
unwrapped all lines to 160 char pep8 standard

wilvk committed:
refactored first level keys of default cache into separate function

wilvk committed:
removed unnecessary brackets

wilvk committed:
removed blank line

wilvk committed:
removed explicit defaults from parameter declaration

wilvk committed:
boto3 check

wilvk committed:
changed tags from list to dict and updated documentation

wilvk committed:
added to doc for tags

wilvk committed:
doc fixes

wilvk committed:
syntax cleanup

wilvk committed:
punctuation, refactoring, fixed bug with tags

wilvk committed:
refactored out common patterns into methods for better readability, fixed bug with default value variable incorrect

wilvk committed:
variable rename

wilvk committed:
minor tweaks

wilvk committed:
minor fixes, changed poll time to 10 seconds, refactoring

wilvk commented:
@s-hertel is there any way I can prevent the shippable module import error from being thrown on importing cloudfront_facts ? Is there an override file for this somewhere?

e.g.

`lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:529:0: ModuleNotFoundError: No module named 'ansible.modules'`

s-hertel commented:
@wilvk I'm not sure how to fix it. The problem with it is that ansible.modules may not be on the managed node.
Running
```
  tasks:
    - name: test import
      cloudfront_distribution:
        profile: shertel
        state: present
        default_origin_domain_name: www.my-cloudfront-origin.com
        tags:
          Name: example distribution
          Project: example project
          Priority: '1'
```
on localhost completes successfully but when I try to run on an ec2 instance I get:
```
fatal: [ec2-34-201-8-213.compute-1.amazonaws.com]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Shared connection to ec2-34-201-8-213.compute-1.amazonaws.com closed.\r\n",
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_k8cmUV/ansible_module_cloudfront_distribution.py\", line 539, in <module>\r\n    from ansible.modules.cloud.amazon.cloudfront_facts import CloudFrontFactsServiceManager\r\nImportError: No module named modules.cloud.amazon.cloudfront_facts\r\n",
    "msg": "MODULE FAILURE",
    "rc": 0
}

PLAY RECAP ********************************************************************************************
ec2-34-201-8-213.compute-1.amazonaws.com : ok=0    changed=0    unreachable=0    failed=1
```

s-hertel commented:
@wilvk I'm not sure how to fix. You could add that class to module_utils for cloudfront and then import it into cloudfront_facts instead of defining it there. The problem with it is that ansible.modules may not be on the managed node.
Running
```
  tasks:
    - name: test import
      cloudfront_distribution:
        profile: shertel
        state: present
        default_origin_domain_name: www.my-cloudfront-origin.com
        tags:
          Name: example distribution
          Project: example project
          Priority: '1'
```
on localhost completes successfully but when I try to run on an ec2 instance I get:
```
fatal: [ec2-34-201-8-213.compute-1.amazonaws.com]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Shared connection to ec2-34-201-8-213.compute-1.amazonaws.com closed.\r\n",
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_k8cmUV/ansible_module_cloudfront_distribution.py\", line 539, in <module>\r\n    from ansible.modules.cloud.amazon.cloudfront_facts import CloudFrontFactsServiceManager\r\nImportError: No module named modules.cloud.amazon.cloudfront_facts\r\n",
    "msg": "MODULE FAILURE",
    "rc": 0
}

PLAY RECAP ********************************************************************************************
ec2-34-201-8-213.compute-1.amazonaws.com : ok=0    changed=0    unreachable=0    failed=1
```

wilvk commented:
@s-hertel thanks for this. I actually asked on the devel mailing group the best way to resolve this and was advised the same thing - place the relevant methods from `cloudfront_facts` into `module_utils`.

Via Adrian Likins:

> For cases like this, the current approach is to move shared code into a ansible.module_utils module and update cloudfront_facts and cloudfront_distribution
> to use the new module_utils.
> 
> Importing 'from ansible.modules.other_module' from my_module isnt recommended, since 'other_module' wont be available when my_module is invoked 
> remotely. The runtime python path for the remote module more or less only has my_module and ansible.module_utils available to it.  
> 

 However, am not sure if it should go in `module_utils/aws` or `module_utils/facts`? Both seem relevant..

wilvk committed:
moved to module_utils/aws

jctanner commented:
bot_skip

wilvk committed:
unwrapped lines

wilvk committed:
moved CloudfrontFactsServiceManager to module_utils/aws as a common class

wilvk committed:
refactoring, fixed bug with add_key_else_validate

wilvk commented:
@s-hertel I've moved the `cloudfront_facts` class to the `module_utils/aws` path and all appears to be working, however, am now getting Shippable errors with trying to find botocore, which seems odd as this was working previously.

See below for an example of the Shippable output:

```
2017-07-30 10:17:11 ERROR: Found 1 import issue(s) on python 3.6 which need to be resolved:
2017-07-30 10:17:11 ERROR: lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:541:0: ModuleNotFoundError: No module named 'botocore' (100%)
2017-07-30 10:17:11 ERROR: The 1 sanity test(s) listed below (out of 2) failed. See error output above for details.
```

wilvk committed:
pep8 fixes

wilvk committed:
pep8 fixes

wilvk commented:
ready_for_review

wilvk committed:
fixes for shippable. copied missing function back to cloudfront_facts, removed from legacy-files, fix for dependency on module in shared resource

wilvk committed:
fixed missing import name

wilvk commented:
ready_for_review

wilvk committed:
removed unnecessary try/except

wilvk committed:
shippable

wilvk committed:
removed unnecessary param validations and try/except

wilvk committed:
error message wording, spacing

wilvk commented:
@jctanner is this PR triggering review requests with the bot skip tag? I requested a review a week ago with no response..

s-hertel commented:
@wilvk I'm sorry, I've seen your request, just have too many things on my plate. I'll get back to this as soon as possible.

willthames commented:
I'd like to get this PR back on track as this would be a valuable module.

There needs to be a separation of the cloudfront_facts module from the cloudfront_distribution module.

Most of the shared `module_utils.aws.cloudformation` functionality is a reimplementation of `snake_dict_to_camel_dict` and `camel_dict_to_snake_dict`. If we need `camel_to_snake` or `snake_to_camel`, let's change `module_utils.ec2`

I'm not a huge fan of classes for modules (as in `module_utils.aws.cloudfront_facts` here) unless it's to provide an interface to multiple implementations (they were arguably useful for moving from rds to rds_v2).

As such, my way forward would be:
* One PR for cloudfront_facts, without the class
* One PR for cloudfront_distribution, without most of the shared library code

It might be worth having a combined shared library for all cloudfront modules, as `module_utils.aws.cloudfront`. I'd hope cloudfront_facts would be the easiest of the PRs to get through, so I'd start with that there.

@wilvk, let me know if that seems like a reasonable proposal and, if so, if you have the time to do it - if not, I can start looking at the improvements based on your existing work.

wilvk commented:
@willthames thanks for the feedback on this one - was a bit stuck on it. I'll have some time this weekend to look at it (still working on the ec2_ami tests stuff) otherwise feel free to start making the changes.

gundalow commented:
As agreed in 2017-12-07 Core meeting, we've done bulk pep8 fixes across all modules in `devel`.

This was done in https://github.com/ansible/ansible/pull/33682 and https://github.com/ansible/ansible/pull/33690 which have been merged into `devel`. because of this you will need to rebase your PR. Thanks in advance.

ryansb commented:
Closing in favor of #31284
